File: drills/ts/solutions/nc50_number_of_one_bits.ts
export function hammingWeight(n: number): number {
  // n & (n - 1) clears the lowest set bit and nothing else, so the loop runs
  // once per one bit rather than once per bit position.
  let count = 0;
  while (n !== 0) {
    n &= n - 1;
    count++;
  }
  return count;
}

File: drills/elixir/solutions/nc50_number_of_one_bits.exs
defmodule Solution do
  import Bitwise

  # n &&& (n - 1) clears the lowest set bit and nothing else, so the recursion
  # runs once per one bit rather than once per bit position.
  def hamming_weight(0), do: 0
  def hamming_weight(n), do: 1 + hamming_weight(n &&& (n - 1))
end
